pallets/pallet-asset-swap/src/traits.rs

```diff
@@ -0,0 +1,75 @@
+// KILT Blockchain – https://botlabs.org
+// Copyright (C) 2019-2024 BOTLabs GmbH
+
+// The KILT Blockchain is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The KILT Blockchain is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+// If you feel like getting in touch with us, you can do so at info@botlabs.org
+
+use xcm::VersionedMultiLocation;
+
+use crate::{Config, LocalCurrencyBalanceOf};
+
+pub trait SwapHooks<T>
+where
+	T: Config,
+{
+	type Error: Into<u8>;
+
+	fn pre_local_to_remote_swap(
+		from: &T::AccountId,
+		to: &VersionedMultiLocation,
+		amount: LocalCurrencyBalanceOf<T>,
+	) -> Result<(), Self::Error>;
+
+	fn post_local_to_remote_swap(
+		from: &T::AccountId,
+		to: &VersionedMultiLocation,
+		amount: LocalCurrencyBalanceOf<T>,
+	) -> Result<(), Self::Error>;
+
+	fn pre_remote_to_local_swap(to: &T::AccountId, amount: u128) -> Result<(), Self::Error>;
+
+	fn post_remote_to_local_swap(to: &T::AccountId, amount: u128) -> Result<(), Self::Error>;
```

Comment:
Why is the `from` not important here?

Reply:
Because unfortunately we don't have information about that. What we get from AH is "deposit these assets into this account", nothing else.

Reply:
So we also don't know, from which parachain the message is coming?

Reply:
We only know that in the `Barrier`. After that, the info is cleared with the `ClearOrigin` instruction. This will change in the future, where the `ClearOrigin` instruction will most likely be replaced by a `DescendOrigin` instruction that the source chain adds to the message, so that we know the `MultiLocation` on the source chain that "authored" the XCM message, but right now the only place where the information about the source chain is available is in `Barrier`s, and that is unfortunately way too early to draw any kind of conclusions or call into any hooks.

Reply:
Okay, but what would happen if a runtime has two instances of the pallet with a different `reserved_location`? The XCM config would use the first struct which is implementing the `swapHook` trait and maybe doing the wrong checks or not?

Reply:
The `SwapHook` trait takes also the pallet instance as a generic parameter, so if the first pallet has a different reserve location, the trait implementation would be able to tell the difference. It would then move on to the second implementation, which would then return the expected result.

Reply:
I see